### PR TITLE
tests: cleanup various uc20 boot tests from previous PR

### PR DESCRIPTION
**Pull request by anonymouse64**
This adds a unit test requested in https://github.com/snapcore/snapd/pull/8228 and refactors how the uc20 uboot image test mocks a uboot bootloader by using bootloader types in bootloadertest to match what we are doing for the other types of mock bootloader.

Also fix a typo.

### CODE REVIEW

**Comment by anonymouse64** [comment]
Also this PR now makes the change to always set `snapd_recovery_system` on the root bootloader on ubuntu-seed, also as suggested in PR #8228

File: boot/makebootable.go [hunk 1]
```diff
@@ -167,6 +167,22 @@ func makeBootable20(model *asserts.Model, rootdir string, bootWith *BootableSet)
...
+	// record which recovery system is to be used on the bootloader, note that
+	// this goes on the main bootloader, and not on the recovery system 
+	// bootloader, for example for grub bootloader, this env var is set on 
+	// the ubuntu-seed root grubenv, and not on the recovery system grubenv in 
+	// the systems/20200314/ subdir on ubuntu-seed
+	blVars := map[string]string{
+		"snapd_recovery_system": bootWith.RecoverySystemLabel,
+	}
+	if err := bl.SetBootVars(blVars); err != nil {
+		return fmt.Errorf("cannot set system environment: %v", err)
```
Comment (pedronis) [inline boot/makebootable.go]: s/system/recovery/

File: boot/makebootable.go [hunk 2]
```diff
@@ -186,12 +202,7 @@ func makeBootable20(model *asserts.Model, rootdir string, bootWith *BootableSet)
 			return fmt.Errorf("cannot extract recovery system kernel assets: %v", err)
 		}
 
-		// record which recovery system is to be used
-		// TODO:UC20: should we do this for all bootloaders?
-		whereRecoveryEnv := map[string]string{
-			"snapd_recovery_system": bootWith.RecoverySystemLabel,
-		}
-		if err := bl.SetBootVars(whereRecoveryEnv); err != nil {
+		if err := bl.SetBootVars(blVars); err != nil {
```
Comment (pedronis) [inline boot/makebootable.go]: I think this is just redoing what we did above and can go away now?

File: boot/makebootable.go (hunk 1 above)
Comment (pedronis) [reply boot/makebootable.go]: relatedly, there's an error in makeBootable20RunMode that says:

"cannot set recovery system environment"

it should drop the "system" bit

File: boot/makebootable_test.go [hunk 3]
```diff
@@ -260,13 +276,43 @@ version: 5.0
...
+	label := "20191209"
+	recoverySystemDir := filepath.Join("/systems", label)
+	bootWith := &boot.BootableSet{
+		RecoverySystemDir: recoverySystemDir,
+		UnpackedGadgetDir: unpackedGadgetDir,
+		Recovery:          true,
+	}
+
+	err = boot.MakeBootable(model, rootdir, bootWith)
+	c.Assert(err, ErrorMatches, "internal error: recovery system label unset")
```
Comment (pedronis) [inline boot/makebootable_test.go]: are the test names and the content matching?

File: boot/makebootable_test.go [hunk 4]
```diff
@@ -374,3 +420,74 @@ base=core20_3.snap
...
+	c.Assert(err, IsNil)
+
+	// ensure only a single file got copied (the uboot.env)
+	files, err := filepath.Glob(filepath.Join(rootdir, "boot/uboot/*"))
+	c.Assert(err, IsNil)
+	c.Check(files, HasLen, 1)
+	// check that the recovery bootloader configuration was copied with
+	// the correct content
+	c.Check(filepath.Join(rootdir, "boot/uboot/uboot.env"), testutil.FileEquals, ubootEnv)
+
```
Comment (pedronis) [inline boot/makebootable_test.go]: shouldn't we check that the label was set in the environment?

File: boot/makebootable.go [hunk 5]
```diff
@@ -167,6 +167,22 @@ func makeBootable20(model *asserts.Model, rootdir string, bootWith *BootableSet)
 		return fmt.Errorf("internal error: cannot find bootloader: %v", err)
 	}
 
+	if bootWith.RecoverySystemLabel == "" {
+		return fmt.Errorf("internal error: recovery system label unset")
```
Comment (pedronis) [inline boot/makebootable.go]: this should probably go as one of the first things in the function

File: boot/makebootable.go (hunk 2 above)
Comment (anonymouse64) [reply boot/makebootable.go]: ah yes I missed this, you're right

File: boot/makebootable_test.go (hunk 3 above)
Comment (anonymouse64) [reply boot/makebootable_test.go]: hmm yes it seems I accidentally swapped test names here

File: tests/main/prepare-image-uboot-uc20/task.yaml [hunk 6]
```diff
@@ -60,7 +60,7 @@ execute: |
 
     echo Running prepare-image as a user
     # TODO:UC20: remove --snap pi_20 once the gadget has been published
-    su -c "SNAPPY_USE_STAGING_STORE=$SNAPPY_USE_STAGING_STORE snap prepare-image --channel edge --snap $ROOT/pi_20-1_arm64.snap $ROOT/model.assertion $ROOT" test
+    session-tool -u test "SNAPPY_USE_STAGING_STORE=$SNAPPY_USE_STAGING_STORE snap prepare-image --channel edge --snap $ROOT/pi_20-1_arm64.snap $ROOT/model.assertion $ROOT"
```
Comment (pedronis) [inline tests/main/prepare-image-uboot-uc20/task.yaml]: this somehow breaks things

File: boot/makebootable.go [hunk 7]
```diff
@@ -167,6 +172,18 @@ func makeBootable20(model *asserts.Model, rootdir string, bootWith *BootableSet)
 		return fmt.Errorf("internal error: cannot find bootloader: %v", err)
 	}
 
+	// record which recovery system is to be used on the bootloader, note that
+	// this goes on the main bootloader, and not on the recovery system 
+	// bootloader, for example for grub bootloader, this env var is set on 
+	// the ubuntu-seed root grubenv, and not on the recovery system grubenv in 
+	// the systems/20200314/ subdir on ubuntu-seed
+	blVars := map[string]string{
```
Comment (bboozzoo) [inline boot/makebootable.go]: Thanks for the change!

**Pull request merged**